### ISSUE TITLE
Train Stucked

### ISSUE THREAD

**Issue by YinengXiong** [vacancy/Synchronized-BatchNorm-PyTorch#44]
Hi ~
I Use
`
if torch.cuda.device_count() > 1:
    model = torch.nn.DataParallel(model)
    model = bnconvert(model)
model.cuda()
`
to use sync-bn during multi-gpu training, but when training the network, it looks like training procedure stucked at final batch in one epoch

![image](https://user-images.githubusercontent.com/23168598/112822689-d8c46b00-90ba-11eb-8d5a-9deba7d5edd3.png)

**Comment by vacancy** [comment]
The implementation requires that each module on different devices should invoke the batchnorm for exactly SAME amount of times in each forward pass. For example, you can not only call batchnorm on GPU0 but not on GPU1. The #i (i = 1, 2, 3, ...) calls of the batchnorm on each device will be viewed as a whole and the statistics will be reduced. This is tricky but is a good way to handle PyTorch's dynamic computation graph. Although sounds complicated, this will usually not be the issue for most of the models.

Can you check this?

**Comment by YinengXiong** [comment]
So if I want to use model e.g. torchvision, which built with nn.BatchNorm
I should: 
0. model = torchvision.models.resnet50()
1. model = bnconvert(model)
2. model = DataParallelWithCallback(model)
3. model.cuda()

Am I right?

**Comment by vacancy** [comment]
Correct. I suspect the reason is the following: 


> The implementation requires that each module on different devices should invoke the batchnorm for exactly SAME amount of times in each forward pass. For example, you can not only call batchnorm on GPU0 but not on GPU1. The #i (i = 1, 2, 3, ...) calls of the batchnorm on each device will be viewed as a whole and the statistics will be reduced. This is tricky but is a good way to handle PyTorch's dynamic computation graph. Although sounds complicated, this will usually not be the issue for most of the models.

**Comment by YinengXiong** [comment]
> Correct. I suspect the reason is the following: 
> 
> 
> > The implementation requires that each module on different devices should invoke the batchnorm for exactly SAME amount of times in each forward pass. For example, you can not only call batchnorm on GPU0 but not on GPU1. The #i (i = 1, 2, 3, ...) calls of the batchnorm on each device will be viewed as a whole and the statistics will be reduced. This is tricky but is a good way to handle PyTorch's dynamic computation graph. Although sounds complicated, this will usually not be the issue for most of the models.
> 

thanks a lot
